allennlp/models/archival.py

```diff
@@ -148,14 +126,6 @@ def archive_model(
         archive.add(weights_file, arcname=_WEIGHTS_NAME)
         archive.add(os.path.join(serialization_dir, "vocabulary"), arcname="vocabulary")
 
-        # If there are supplemental files to archive:
-        if files_to_archive:
```

Comment:
How does `cached_path` handle this situation? That is, some model uses ELMo on a downstream task and fine tunes the ELMo weights. This code will ensure that the fine-tuned weights are saved. This is a very common use case, no? Certainly `cached_path` works in the read-only case. Sorry if I'm missing something obvious...

Reply:
Actually, sorry. Let me look at this more closely.

Reply:
The model always has its weights that are learned / fine-tuned in `weights.th`, the main component of our `model.tar.gz` files.  When using ELMo, the actual ELMo weights always end up as part of the model, because it's part of some `TokenEmbedder` that the model keeps as a `Module`.  The "weight_file" parameter that I removed from the file archive [here](https://github.com/allenai/allennlp/pull/3674/files#diff-941a4360ef08efef82e2b72210a7ab93L228) only has the *original* ELMo weights.  It's only actually used at initialization, and gets immediately overwritten when you load the saved weights in the model archive (this is an annoying aspect of how our model loading works, which I don't know how to get around) (also note that whether you fine-tuned the ELMo weights or not doesn't matter here).  There's also an "options_file" in the line above the one that I linked, and the contents of that one is more important at model loading time - it defines sizes and things, and without it we couldn't load a model.  But we can just specify a URL to that file, instead of a local path, and everything works exactly as it does now, without the need for `files_to_archive` and the accompanying `from_params` logic.

Reply:
On a more careful reading it now looks to me like with the old code (in the fine-tuning ELMo case) we would copy the ELMo weight file (before fine-tuning!) into the archive. Then, when loading, we would load those files while constructing the `Elmo` module only to overwrite them with the actually fine-tuned weights stored separately in the archive. Is that correct?

Reply:
We need a mutex on these comments ;)